api/include/opentelemetry/trace/tracer_provider.h

```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/trace/tracer.h"
+
+namespace opentelemetry
+{
+namespace trace
+{
+class TracerProvider
+{
+public:
+  virtual ~TracerProvider()                                                 = default;
+  virtual Tracer *const GetTracer(nostd::string_view library_name,
+                                  nostd::string_view library_version,
+                                  const TracerOptions &tracer_options = {}) = 0;
```

Comment:
Could we have this instead return a type-erased shared_ptr (something similar to [this](https://gist.github.com/rnburn/886f677c79a43df64fa95b495a368524))?

That would still provide stability across different STL implementations and then we don't have to worry about a TracerProvider getting destroyed while tracers are still in use.

Reply:
I'm thinking this should go into a follow up PR to get this monster closed. Thoughts?

Reply:
I'd rather not merge something that introduces a lifetime issue (even if it's only a placeholder).

What about adding a separate PR for a type-erased shared_ptr? then updating this PR when it gets merged